crates/miden-objects/src/account/component/metadata/storage/type_registry.rs

```diff
@@ -0,0 +1,615 @@
+
+/// Trait for converting a string into a single `Felt`.
+pub trait FeltType: Send + Sync {
+    /// Returns the type identifier.
+    fn type_name() -> SchemaTypeIdentifier
+    where
+        Self: Sized;
+
+    /// Parses the input string into a `Felt`.
+    fn parse_felt(input: &str) -> Result<Felt, SchemaTypeError>
+    where
+        Self: Sized;
+
+    /// Displays a `Felt` in a canonical string representation for this type.
+    fn display_felt(value: Felt) -> Result<String, SchemaTypeError>
+    where
+        Self: Sized;
+}
+
+/// Trait for converting a string into a single `Word`.
+pub trait WordType: alloc::fmt::Debug + Send + Sync {
+    /// Returns the type identifier.
+    fn type_name() -> SchemaTypeIdentifier
+    where
+        Self: Sized;
+
+    /// Parses the input string into a `Word`.
+    fn parse_word(input: &str) -> Result<Word, SchemaTypeError>
+    where
+        Self: Sized;
+
+    /// Displays a `Word` in a canonical string representation for this type.
+    fn display_word(value: Word) -> Result<String, SchemaTypeError>
+    where
+        Self: Sized;
+}
+
+impl<T> WordType for T
+where
+    T: FeltType + alloc::fmt::Debug,
+{
+    fn type_name() -> SchemaTypeIdentifier {
+        <T as FeltType>::type_name()
+    }
+
+    fn parse_word(input: &str) -> Result<Word, SchemaTypeError> {
+        let felt = <T as FeltType>::parse_felt(input)?;
+        Ok(Word::from([Felt::new(0), Felt::new(0), Felt::new(0), felt]))
+    }
+
+    fn display_word(value: Word) -> Result<String, SchemaTypeError> {
+        if value[0] != Felt::new(0) || value[1] != Felt::new(0) || value[2] != Felt::new(0) {
+            return Err(SchemaTypeError::ConversionError(format!(
+                "expected a word of the form [0, 0, 0, <felt>] for type `{}`",
+                Self::type_name()
+            )));
+        }
+        <T as FeltType>::display_felt(value[3])
+    }
+}
+
+// FELT IMPLS FOR NATIVE TYPES
+// ================================================================================================
+
+/// A felt type that represents irrelevant elements in a storage schema definition.
+struct Void;
+
+impl FeltType for Void {
+    fn type_name() -> SchemaTypeIdentifier {
+        SchemaTypeIdentifier::void()
+    }
+
+    fn parse_felt(input: &str) -> Result<Felt, SchemaTypeError> {
+        let parsed = <Felt as FeltType>::parse_felt(input)?;
+        if parsed != Felt::new(0) {
+            return Err(SchemaTypeError::ConversionError("void values must be zero".to_string()));
+        }
+        Ok(Felt::new(0))
+    }
+
+    fn display_felt(value: Felt) -> Result<String, SchemaTypeError> {
+        if value != Felt::new(0) {
+            return Err(SchemaTypeError::ConversionError("void values must be zero".to_string()));
+        }
+        Ok("0".into())
+    }
+}
+
+impl FeltType for u8 {
+    fn type_name() -> SchemaTypeIdentifier {
+        SchemaTypeIdentifier::new("u8").expect("type is well formed")
```

Comment:
Nit: These are also native types, right? Should we also have `SchemaTypeIdentifier::u8()`, etc.?